src/comm/Connection.h

```diff
@@ -137,12 +140,9 @@ class Connection: public CodeContext
     virtual ScopedId codeContextGist() const override;
     virtual std::ostream &detailCodeContext(std::ostream &os) const override;
 
-private:
-    /** These objects may not be exactly duplicated. Use copyDetails() instead. */
-    Connection(const Connection &c);
-
-    /** These objects may not be exactly duplicated. Use copyDetails() instead. */
-    Connection & operator =(const Connection &c);
+    // TODO: Declare and use (here and elsewhere) NonCopyable, NonMovable, etc.
+    /// no C++ copying/moving; see clone*() methods for custom/partial copies
+    Connection &operator =(Connection &&) = delete;
```

Comment:
FTR: The explanation behind this change is in the last paragraph of the unofficial commit 46f70173f1ab392bf3b8b6755f0fdfab8124e1e2 message, quoted here:

> I had to refactor undefined Connection copy constructors because their comments referred to the renamed copyDetails() and should now refer to the second cloning method as well. Modern code should also avoid making these methods private and should disable moving (where it is unsupported). Deleting the move assignment operator is not consistent with how we often delete the move constructor instead, but I found some references that seem to recommend this method now, including https://stackoverflow.com/a/48865077. I left a TODO that would refactor this properly, for all affected classes.

@kinkie, BTW, this may be a good TODO candidate for you: It is relatively compact, useful for future development, and arguably educational/interesting (if a bit challenging). The key for this TODO is to come up with the smallest set of reusable base classes that leaves no holes in the current (and clearly foreseeable future) Squid needs. Taking a clue from existing libraries that do this already (e.g., Boost?) may be a good idea, but blind copying of their APIs would probably miss some important Squid specifics. Locating existing Squid classes that should get these base classes instead of their ad hoc hacks may be a bit challenging as well.

Reply:
FTR: Please do ensure that all non-copy + non-move objects related to logged state data have a Pointer which can be copied and moved around as needed. Otherwise we end up with performance regressions for them.


@rousskov, why is move now being forbidden? move is completely safe for this object. Copy is only forbidden due to FD close races which are impossible for move semantics.

Reply:
> why is move now being forbidden? 

The question implies that the move was allowed before this PR but is now forbidden. This PR does not change the Connection moving ban: It was forbidden before and it remains forbidden now. This PR accomplishes that in one line of code (instead of three), covers more cases at compile time (instead of linking time), and leads to better compiler error messages. Other than those improvements, there are no material changes here. <!-- https://godbolt.org/z/hHAUvg -->

The copy/move-banning code will become clearer when the TODO (affecting all such cases) is implemented. The same could have been said about the old code as well, of course.

> move is completely safe for this object.

I do not know what you mean exactly, but the default/generated move operations would be deadly for Connection (before and after these PR changes). This statement is true for nearly any object with a non-default destructor. In this particular case, that destructor will close the underlying connection (if it was open), leaving the moved copy with a stale `fd` value (among other problems).

> Copy is only forbidden due to FD close races which are impossible for move semantics.

Custom code is required to implement that move semantics. Connection lacks that code. While one could correctly implement a custom move constructor and move assignment operator for Connection, there is currently no good reason to do that because there is no code that would use that optimization. That may change, of course.

BTW, that custom moving code will have to overcome more obstacles than FD ownership, but that is not important here.

Reply:
The old code explicitly forbid (and documented) copy as not allowed. Leaving move up to whatever the compiler chose.

The new code says nothing about copy and forbids move. That is wrong since copy is the bad operation, move we should not care about.

Reply:
I am sorry, but your statements contradict how C++ works. There is no "compiler choice" here now and there was not one before this change. The same things were and are prohibited/allowed.

Reply:
@eduard-bagdasaryan, to make progress in these sad circumstances, I suggest restoring the old `private` undefined method declarations and just updating their duplicated comments to reference the new `clone*()` methods names. And drop the TODO, just in case!

Reply:
Done (dfc20a2).

Reply:
Thank you!